polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1343,6 +1343,119 @@ class RaiseSetNDRange
     Value value;
   };
 };
+
+class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
+public:
+  using OpHostRaisePattern<LLVM::VarAnnotation>::OpHostRaisePattern;
+
+  LogicalResult matchAndRewrite(LLVM::VarAnnotation op,
+                                PatternRewriter &rewriter) const final {
+    // The annotation must be inside a handler function.
+    auto enclosingFunc = op->getParentOfType<LLVM::LLVMFuncOp>();
+    if (!enclosingFunc || !getHandler(enclosingFunc))
+      return failure();
+
+    // It must be a "kernel" annotation.
+    FailureOr<StringRef> failureOrAnnotationStr = getAnnotation(op);
+    if (failed(failureOrAnnotationStr) || *failureOrAnnotationStr != "kernel")
+      return failure();
+
+    // There should be exactly one store to the marked pointer.
+    Value annotatedPtr = op.getVal();
+    auto store = getUniqueStore(annotatedPtr);
+    if (!store)
+      return failure();
+
+    Value lambdaObj = store.getValue();
+    auto alloca = dyn_cast_or_null<LLVM::AllocaOp>(lambdaObj.getDefiningOp());
+    if (!alloca || !alloca.getElemType().has_value())
+      return failure();
+
+    auto lambdaClassTy =
+        dyn_cast<LLVM::LLVMStructType>(alloca.getElemType().value());
+    if (!lambdaClassTy)
+      return failure();
+
+    auto captureTypes = lambdaClassTy.getBody();
+
+    // Look for stores that represent the capturing of kernel args.
+    // Assuming the no-op `getelementptr %lambdaObj[0, 0]` was folded, the
+    // first capture is actually a store to the pointer.
+    if (auto capture = getUniqueStore(lambdaObj)) {
+      Value captured = tryToBroaden(capture.getValue(), captureTypes[0]);
+      rewriter.setInsertionPointAfter(capture);
+      rewriter.create<sycl::SYCLHostSetCaptured>(capture->getLoc(), lambdaObj,
```

Comment:
I'm proposing to use `lambdaObj` here (which is typically an `alloca` of the capture-specific struct type) rather than `annotatedPtr` (a bare ptr alloca which `lambdaObj` is copied into).

Given that neither one is directly used by a function call of interest, we'll have to do some form of copy-propagation when raising the `launch_kernel` op anyways, hence I thought it make sense to use the more specific value for the `set_captured` op.